ops/testing.py

```diff
@@ -1804,8 +1865,28 @@ def action_log(self, message):  # type:ignore
     def action_fail(self, message=''):  # type:ignore
         raise NotImplementedError(self.action_fail)  # type:ignore
 
-    def network_get(self, endpoint_name, relation_id=None):  # type:ignore
-        raise NotImplementedError(self.network_get)  # type:ignore
+    def network_get(self, binding_name: str, relation_id: Optional[int] = None) -> '_NetworkDict':
+        if relation_id is None:
+            data = self._networks.get((binding_name, None))
+            if data is not None:
+                # Get the default data if it exists
+                return data
+            # Otherwise get the data with the lowest relation_id as the default
+            # TODO: does this even make sense?
+            datas = sorted((r, data) for (b, r), data in self._networks.items()
+                           if b == binding_name)
+            if datas:
+                return datas[0][1]
+        else:
+            data = self._networks.get((binding_name, relation_id))
+            if data is not None:
+                # Get the data for this specific relation_id if it exists
+                return data
+            data = self._networks.get((binding_name, None))
```

Comment:
you already have a branch for relation_id == None at line 1868, maybe good to raise this up a bit and consolidate the logic?

Reply:
This has gone away with John's suggested approach.